tsnet/tsnet.go

```diff
@@ -63,12 +63,23 @@ type Server struct {
 
 	mu        sync.Mutex
 	listeners map[listenKey]*listener
+	dialer    *tsdial.Dialer
 }
 
-func (s *Server) doInit() {
-	if err := s.start(); err != nil {
-		s.initErr = fmt.Errorf("tsnet: %w", err)
+func (s *Server) Dial(ctx context.Context, network, address string) (net.Conn, error) {
+	if err := s.init(); err != nil {
+		return nil, err
 	}
+	return s.dialer.UserDial(ctx, network, address)
+}
+
+func (s *Server) init() error {
+	s.initOnce.Do(func() {
```

Comment:
revert this back to a `doInit` method so this closure doesn't allocate, closing over `s.